aspnetcore/fundamentals/http-context.md

```diff
@@ -0,0 +1,146 @@
+
+The Razor Pages `PageModel` exposes the `HttpContext` property:
+
+```csharp
+public class AboutModel : PageModel
+{
+    public string Message { get; set; }
+
+    public void OnGet()
+    {
+        Message = HttpContext.Request.PathBase;
+    }
+}
+```
+
+Controllers expose the `HttpContext` property from the `ControllerBase`:
+
+```csharp
+public IActionResult About()
+{
+    var pathBase = HttpContext.Request.PathBase;
+    // Do something with the PathBase.
+
+    return View();
+}
+```
+
+When working with custom middleware components, `HttpContext` is passed into the `Invoke` method and can be accessed when the middleware is being configured.
+
+```csharp
+public class MyCustomMiddleware
+{
+    public async Task Invoke(HttpContext context)
+    {
+        // Middleware initialization optionally using HttpContext
+    }
+    ...
+}
+```
+
+## Use HttpContext from custom components
+
+For other framework and custom components that need access to `HttpContext`, the recommended approach is to register a dependency using the built-in [dependency injection](xref:fundamentals/dependency-injection) container. The dependency injection container will supply the `IHttpContextAccessor` to any classes that declare it as a dependency in their constructors.
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+     services.AddMvc();
+     services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
+     services.AddTransient<IUserRepository, UserRepository>();
+}
+```
+
+In the preceding example:
+
+* `UserRepository` declares its dependency on `IHttpContextAccessor`.
+* The dependency is supplied when dependency injection resolves the dependency chain and creates an instance of `UserRepository`.
+
+```csharp
+public class UserRepository : IUserRepository
+{
+    private readonly IHttpContextAccessor _httpContextAccessor;
+
+    public UserRepository(IHttpContextAccessor httpContextAccessor)
+    {
+        _httpContextAccessor = httpContextAccessor;
+    }
+
+    public void LogCurrentUser()
+    {
+        var username = _httpContextAccessor.HttpContext.User.Identity.Name;
+        service.LogAccessRequest(username);
+    }
+
+    ...
+```
+
+## Work with libraries and ported code
+
+Registering IHttpContextAccessor in the application startup using the built-in dependency injection framework is the preferred approach to accessing HttpContext. In scenarios where it is not feasible to refactor to this approach, it is possible to simulate the behavior of `System.Web.HttpContext` from ASP.NET 4.x.
+
+Register `IHttpContextAccessor` as a dependency in the `ConfigureServices` method.
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
+}
+```
+
+Create a wrapper class for `HttpContext` and provide a method to invoke to pass an instance of `IHttpContextAccessor`.
```

Comment:
I would remove this entire section. I don't want this recommendation in our official documentation.

Reply:
Is there another place you think it would be more appropriate? Webdev blog? We've worked with customers recently who have had trouble porting large ASP.NET projects in part because of HttpContext.Current usage. It seems like this is a useful technique to document somewhere (with clear caveats that it is only recommended as a short-term bridge until ported code can be more properly refactored).

Reply:
Makes sense. Maybe somewhere in the [migration docs](https://docs.microsoft.com/en-us/aspnet/core/migration/proper-to-2x/)?

Reply:
Yep!

Reply:
@coderandhiker Want to pull this section from here so the IHttpContextAccessor docs just focus on the recommended usage in new development and then make a follow-up PR to add a paragraph/page/whatever to the migration docs covering how this can help porting old 'HttpContext.Current' usage?